packages/dataviews/src/components/dataviews/stories/index.story.tsx

```diff
@@ -467,32 +472,108 @@ export const InfiniteScroll = () => {
 		email?: string;
 	} ) => item.id.toString();
 
-	const infiniteScrollHandler = useCallback( () => {
-		if ( isLoadingMore || currentPage >= totalPages ) {
-			return;
-		}
+	const infiniteScrollHandler = useCallback(
+		( direction: 'up' | 'down' ) => {
+			if ( isLoadingMore ) {
+				return;
+			}
 
-		setIsLoadingMore( true );
+			// Handle scrolling down to load next page
+			if ( direction === 'down' && currentPage < totalPages ) {
+				setIsLoadingMore( true );
+				setScrollDirection( 'down' );
+				setView( {
+					...view,
+					page: currentPage + 1,
+				} );
+			}
 
-		setView( {
-			...view,
-			page: currentPage + 1,
-		} );
-	}, [ isLoadingMore, currentPage, totalPages, view ] );
+			// Handle scrolling up to load previous page
+			if ( direction === 'up' && currentPage > 1 ) {
+				setIsLoadingMore( true );
+				setScrollDirection( 'up' );
+				setView( {
+					...view,
+					page: currentPage - 1,
+				} );
+			}
+		},
+		[ isLoadingMore, currentPage, totalPages, view ]
+	);
+
+	const intersectionObserverCallback: IntersectionObserverCallback =
+		useCallback( ( entries: IntersectionObserverEntry[] ) => {
+			// Calculate new visible entries outside of setState
+			setVisibleEntries( ( prev ) => {
+				const newVisibleEntries = new Set( prev );
+				let hasChanged = false;
+
+				entries.forEach( ( entry ) => {
+					const posInSet = Number(
```

Comment:
This callback depends on dataviews layout implementation details so I'm not sure it should be defined by the consumer. I'll try to think of a better way of just making the info about visible items available to consumers instead.